src/util/psbt/mod.rs

```diff
@@ -198,6 +206,526 @@ impl PartiallySignedTransaction {
+        if self.signing_algorithm(input_index)? != SigningAlgorithm::Ecdsa {
+            return Err(SignError::WrongSigningAlgorithm);
+        }
+
+        let input = self.checked_input(input_index)?;
+        let utxo = self.spend_utxo(input_index)?;
+        let spk = &utxo.script_pubkey; // scriptPubkey for input spend utxo.
+
+        let hash_ty = input.ecdsa_hash_ty()
+            .map_err(|_| SignError::InvalidSighashType)?; // Only support standard sighash types.
+
+        let output_type = self.output_type(input_index)?;
+
+        let is_wpkh = output_type == OutputType::Wpkh;
+        let is_nested_wpkh = output_type == OutputType::ShWpkh;
+
+        let sighash = if output_type.is_segwitv0() {
+            if is_wpkh || is_nested_wpkh {
+                let script_code = if is_wpkh {
+                    Script::p2wpkh_script_code(spk).ok_or(SignError::NotWpkh)?
+                } else {
+                    Script::p2wpkh_script_code(input.redeem_script.as_ref().expect("checked above"))
+                        .ok_or(SignError::NotWpkh)?
+                };
+                cache.segwit_signature_hash(input_index, &script_code, utxo.value, hash_ty)?
+            } else {
+                let script_code = input.witness_script.as_ref().ok_or(SignError::MissingWitnessScript)?;
+                cache.segwit_signature_hash(input_index, script_code, utxo.value, hash_ty)?
+            }
+        } else {
+            let script_code = if spk.is_p2sh() {
+                input.redeem_script.as_ref().ok_or(SignError::MissingRedeemScript)?
+            } else {
+                spk
+            };
+            cache.legacy_signature_hash(input_index, script_code, hash_ty.to_u32())?
+        };
+
+        Ok((Message::from_slice(&sighash).expect("sighashes are 32 bytes"), hash_ty))
+    }
+
+    /// Returns the spending utxo for this PSBT's input at `input_index`.
+    pub fn spend_utxo(&self, input_index: usize) -> Result<&TxOut, SignError> {
+        let input = self.checked_input(input_index)?;
+        let utxo = if let Some(witness_utxo) = &input.witness_utxo {
+            witness_utxo
+        } else if let Some(non_witness_utxo) = &input.non_witness_utxo {
+            let vout = self.unsigned_tx.input[input_index].previous_output.vout;
+            &non_witness_utxo.output[vout as usize]
+        } else {
+             return Err(SignError::MissingSpendUtxo);
+        };
+        Ok(utxo)
+    }
+
+    /// Gets the input at `input_index` after checking that it is a valid index.
+    fn checked_input(&self, input_index: usize) -> Result<&Input, SignError> {
+        self.check_index_is_within_bounds(input_index)?;
+        Ok(&self.inputs[input_index])
+    }
+
+    /// Checks `input_index` is within bounds for the PSBT `inputs` array and
+    /// for the PSBT `unsigned_tx` `input` array.
+    fn check_index_is_within_bounds(&self, input_index: usize) -> Result<(), SignError> {
```

Comment:
In 3da72f06f8596bdb4322e2e20f3d3ab12fe9491e:

I think we should just `assert!` that the two bounds are the same and then just check one.

Reply:
I don't know what situations lead to these two lengths being different but purely from a code perspective I don't see the benefit of panicing in a function that returns an error and is only there to do the check. Or am I missing something?

Reply:
The benefit is panicking is that if these lengths ever differ then an invariant has been violated and we want to know ASAP.

Reply:
>I think we should just assert! that the two bounds are the same and then just check one.

Do you mean that we do the out-of-bounds check multiple times? So far, we have avoided unnecessary unwraps in function that already returned Result

Reply:
>So far, we have avoided unnecessary unwraps in function that already returned Result

Unnecessary unwraps, perhaps, because `unwrap` increases cognitive load. But here we are indexing, so there is still a panic path, *and* doing an unnecessary range check, *and* missing an opportunity to assert an invariant.

Reply:
We now have
```rust
    fn check_index_is_within_bounds(&self, input_index: usize) -> Result<(), SignError> {
        assert_eq!(self.inputs.len(), self.unsigned_tx.input.len(), "input length invariant violated");

        if input_index >= self.inputs.len() {
            return Err(SignError::IndexOutOfBounds(input_index, self.inputs.len()));
        }

        Ok(())
    }
```